bulbs/recirc/mixins.py

```diff
@@ -0,0 +1,86 @@
+from django.conf import settings
+from django.db import models
+
+from elasticsearch import Elasticsearch
+from json_field import JSONField
+
+from bulbs.content.custom_search import custom_search_model
+from bulbs.content.models import Content
+
+
+es = Elasticsearch(settings.ES_CONNECTIONS["default"]["hosts"])
+
+
+class BaseQueryMixin(models.Model):
+
+    query = JSONField(default={}, blank=True)
+
+    class Meta:
+        abstract = True
+
+    def clean(self):
+        super(BaseQueryMixin, self).clean()
+        self.clean_query()
+
+    def clean_query(self):
+        """
+        Removes any `None` value from an elasticsearch query.
+        """
+        if self.query and self.query != {}:
+            for key, value in self.query.items():
+                if isinstance(value, list) and None in value:
+                    self.query[key] = [v for v in value if v is not None]
+
+    def save(self, *args, **kwargs):
+        self.clean()
+        super(BaseQueryMixin, self).save()
+
+    def get_recirc_content(self, published=True):
+        """gets the first 3 content objects in the `included_ids`
+        """
+        if "query" in self.query:
+            q = self.query["query"]
+        else:
+            q = self.query
+
+        search = custom_search_model(Content, q, published=published, field_map={
+            "feature_type": "feature_type.slug",
+            "tag": "tags.slug",
+            "content-type": "_type"
+        })
+        return search[:3]
+
+    def get_full_recirc_content(self, published=True):
+        """performs es search and gets all content objects
+        """
+        if "query" in self.query:
+            q = self.query["query"]
+        else:
+            q = self.query
+
+        search = custom_search_model(Content, q, published=published, field_map={
+            "feature_type": "feature_type.slug",
+            "tag": "tags.slug",
+            "content-type": "_type"
+        })
+        return search
+
+    @property
+    def contents(self):
+        """performs .get_content() and caches it in a ._content property
+        """
+        if not hasattr(self, "_content"):
+            self._content = self.get_content()
+        return self._content
```

Comment:
This is OK, but we should look into using some kind of memoizing decorator. 

I used Pyramid's `reify` in a past life: https://github.com/Pylons/pyramid/blob/master/pyramid/decorator.py#L4